File: bot.py
#!/usr/bin/env python3
"""Comment-to-DM per @ferroshaolin.

Un passaggio per esecuzione (pensato per launchd/cron ogni ~2 minuti):
 1. legge gli ultimi post dell'account Instagram
 2. cerca nei commenti nuovi le parole chiave di config.json
 3. a chi le scrive manda una private reply (DM) e, se configurata,
    una risposta pubblica sotto il commento
 4. segna in state.json i commenti già gestiti (mai doppioni)

Testi e parole chiave: config.json. Credenziali: .env. Log: bot.log.
"""
import json
import re
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone
from pathlib import Path

BASE = Path(__file__).parent
GRAPH = "https://graph.facebook.com/v23.0"
LOG_PATH = BASE / "bot.log"
STATE_PATH = BASE / "state.json"


def log(msg: str) -> None:
    line = f"{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}  {msg}"
    print(line)
    with LOG_PATH.open("a") as f:
        f.write(line + "\n")


def load_env() -> dict:
    env = {}
    for line in (BASE / ".env").read_text().splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            k, _, v = line.partition("=")
            env[k.strip()] = v.strip()
    return env


def load_json(path: Path, default):
    if path.exists():
        return json.loads(path.read_text())
    return default


def api_get(path: str, params: dict) -> dict:
    url = f"{GRAPH}/{path}?" + urllib.parse.urlencode(params)
    with urllib.request.urlopen(url, timeout=30) as r:
        return json.loads(r.read())


def api_post(path: str, payload: dict, token: str) -> dict:
    url = f"{GRAPH}/{path}?" + urllib.parse.urlencode({"access_token": token})
    req = urllib.request.Request(
        url,
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as r:
        return json.loads(r.read())


def api_error(e: urllib.error.HTTPError) -> str:
    try:
        err = json.loads(e.read()).get("error", {})
        return f"{err.get('code')}/{err.get('error_subcode')}: {err.get('message')}"
    except Exception:
        return f"HTTP {e.code}"


def matches(text: str, keywords: list) -> bool:
    t = text.lower()
    return any(re.search(rf"\b{re.escape(k.lower())}\b", t) for k in keywords)


def main() -> None:
    env = load_env()
    cfg = load_json(BASE / "config.json", None)
    if cfg is None:
        sys.exit("config.json mancante")
    for k in ("PAGE_ID", "PAGE_TOKEN", "IG_USER_ID"):
        if not env.get(k):
            sys.exit(f"{k} mancante nel .env: lanciare prima setup_token.py")

    token = env["PAGE_TOKEN"]
    state = load_json(STATE_PATH, {"replied": {}, "own_username": None})

    if not state.get("own_username"):
        me = api_get(env["IG_USER_ID"], {"fields": "username", "access_token": token})
        state["own_username"] = me.get("username", "")

    cutoff = datetime.now(timezone.utc) - timedelta(days=cfg["max_comment_age_days"])
    handled = 0

    media = api_get(f"{env['IG_USER_ID']}/media", {
        "fields": "id,timestamp",
        "limit": cfg["media_limit"],
        "access_token": token,
    }).get("data", [])

    for m in media:
        comments = api_get(f"{m['id']}/comments", {
            "fields": "id,text,timestamp,username",
            "limit": 50,
            "access_token": token,
        }).get("data", [])

        for c in comments:
            cid = c["id"]
            entry = state["replied"].get(cid)
            if entry and (entry.get("ok") or entry.get("attempts", 0) >= 3):
                continue
            if c.get("username", "") == state["own_username"]:
                continue
            ts = datetime.strptime(c["timestamp"], "%Y-%m-%dT%H:%M:%S%z")
            if ts < cutoff:
                continue
            if not matches(c.get("text", ""), cfg["keywords"]):
                continue

            entry = entry or {"attempts": 0}
            entry["attempts"] += 1
            try:
                api_post(f"{env['PAGE_ID']}/messages", {
                    "recipient": {"comment_id": cid},
                    "message": {"text": cfg["dm_message"]},
                }, token)
                entry["ok"] = True
                log(f"DM inviato a @{c.get('username')} (commento: {c.get('text', '')[:60]!r})")
                handled += 1
            except urllib.error.HTTPError as e:
                entry["error"] = api_error(e)
                log(f"ERRORE DM a @{c.get('username')}: {entry['error']}")

            if entry.get("ok") and cfg.get("public_reply"):
                try:
                    api_post(f"{cid}/replies", {"message": cfg["public_reply"]}, token)
                except urllib.error.HTTPError as e:
                    log(f"ERRORE risposta pubblica a @{c.get('username')}: {api_error(e)}")

            state["replied"][cid] = entry
            STATE_PATH.write_text(json.dumps(state, indent=1))
            time.sleep(1)

    STATE_PATH.write_text(json.dumps(state, indent=1))
    if handled:
        log(f"Passaggio completato: {handled} DM inviati")


if __name__ == "__main__":
    try:
        main()
    except urllib.error.HTTPError as e:
        log(f"ERRORE API: {api_error(e)}")
        sys.exit(1)
    except Exception as e:
        log(f"ERRORE: {e}")
        sys.exit(1)

File: bot.rb
#!/usr/bin/ruby
# Comment-to-DM per @ferroshaolin — versione Ruby (zero dipendenze, gira
# con il Ruby di serie di macOS: niente da installare).
# Un passaggio per esecuzione, pensato per launchd ogni ~2 minuti:
#  1. legge gli ultimi post dell'account Instagram
#  2. cerca nei commenti nuovi le parole chiave delle regole di config.json
#     (la prima regola che combacia decide il DM da mandare)
#  3. manda la private reply (DM) e una risposta pubblica, ruotando tra
#     le varianti di public_replies per sembrare naturale
#  4. segna i commenti gestiti in state.json (mai doppioni)
require "net/http"
require "json"
require "time"
require "uri"

BASE  = File.expand_path(File.dirname(__FILE__))
GRAPH = "https://graph.facebook.com/v23.0"
LOG   = File.join(BASE, "bot.log")
STATE = File.join(BASE, "state.json")

def log(msg)
  line = "#{Time.now.strftime('%Y-%m-%d %H:%M:%S')}  #{msg}"
  puts line
  File.open(LOG, "a") { |f| f.puts line }
end

def load_env
  env = {}
  File.read(File.join(BASE, ".env")).each_line do |line|
    line = line.strip
    next if line.empty? || line.start_with?("#") || !line.include?("=")
    k, v = line.split("=", 2)
    env[k.strip] = v.strip
  end
  env
end

def api_get(path, params)
  uri = URI("#{GRAPH}/#{path}")
  uri.query = URI.encode_www_form(params)
  res = Net::HTTP.get_response(uri)
  body = JSON.parse(res.body)
  raise "API #{res.code}: #{body.dig('error', 'message')}" unless res.is_a?(Net::HTTPSuccess)
  body
end

def api_post(path, payload, token)
  uri = URI("#{GRAPH}/#{path}?#{URI.encode_www_form(access_token: token)}")
  req = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
  req.body = JSON.generate(payload)
  res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |h| h.request(req) }
  body = JSON.parse(res.body) rescue {}
  unless res.is_a?(Net::HTTPSuccess)
    err = body["error"] || {}
    raise "#{err['code']}/#{err['error_subcode']}: #{err['message']}"
  end
  body
end

def matches?(text, keywords)
  t = text.to_s.downcase
  keywords.any? { |k| t =~ /\b#{Regexp.escape(k.downcase)}\b/ }
end

env = load_env
cfg = JSON.parse(File.read(File.join(BASE, "config.json")))
%w[PAGE_ID PAGE_TOKEN IG_USER_ID].each do |k|
  abort "#{k} mancante nel .env" if env[k].to_s.empty?
end
abort "config.json senza 'rules'" unless cfg["rules"].is_a?(Array) && !cfg["rules"].empty?
token = env["PAGE_TOKEN"]
state = File.exist?(STATE) ? JSON.parse(File.read(STATE)) : {}
state["replied"]   ||= {}
state["reply_idx"] ||= 0

if state["own_username"].to_s.empty?
  me = api_get(env["IG_USER_ID"], { "fields" => "username", "access_token" => token })
  state["own_username"] = me["username"].to_s
end

cutoff  = Time.now.utc - cfg["max_comment_age_days"] * 86_400
handled = 0

begin
  media = api_get("#{env['IG_USER_ID']}/media", {
    "fields" => "id,timestamp", "limit" => cfg["media_limit"], "access_token" => token
  })["data"] || []

  media.each do |m|
    comments = api_get("#{m['id']}/comments", {
      "fields" => "id,text,timestamp,username", "limit" => 50, "access_token" => token
    })["data"] || []

    comments.each do |c|
      cid   = c["id"]
      entry = state["replied"][cid]
      next if entry && (entry["ok"] || entry["attempts"].to_i >= 3)
      next if c["username"].to_s == state["own_username"]
      next if Time.iso8601(c["timestamp"]) < cutoff
      rule = cfg["rules"].find { |r| matches?(c["text"], r["keywords"]) }
      next unless rule

      entry ||= { "attempts" => 0 }
      entry["attempts"] += 1
      begin
        api_post("#{env['PAGE_ID']}/messages", {
          "recipient" => { "comment_id" => cid },
          "message"   => { "text" => rule["dm_message"] },
        }, token)
        entry["ok"] = true
        log "DM inviato a @#{c['username']} [#{rule['keywords'].first}] (commento: #{c['text'].to_s[0, 60].inspect})"
        handled += 1
      rescue => e
        entry["error"] = e.message
        log "ERRORE DM a @#{c['username']}: #{e.message}"
      end

      replies = cfg["public_replies"] || []
      if entry["ok"] && !replies.empty?
        reply_text = replies[state["reply_idx"].to_i % replies.size]
        begin
          api_post("#{cid}/replies", { "message" => reply_text }, token)
          state["reply_idx"] = state["reply_idx"].to_i + 1
        rescue => e
          log "ERRORE risposta pubblica a @#{c['username']}: #{e.message}"
        end
      end

      state["replied"][cid] = entry
      File.write(STATE, JSON.pretty_generate(state))
      sleep 1
    end
  end

  # --- DM diretti: chi scrive "info"/"corso" in chat riceve la stessa risposta ---
  # Solo se la config lo abilita (dm_keywords_enabled), una volta per conversazione,
  # solo quando l'ULTIMO messaggio è dell'utente (così non risponde dove Ale sta già
  # scrivendo a mano) e contiene una parola chiave.
  if cfg["dm_keywords_enabled"]
   begin
    state["dm_replied"] ||= {}
    convos = api_get("#{env['IG_USER_ID']}/conversations", {
      "platform"     => "instagram",
      "fields"       => "id,messages.limit(5){id,from,message,created_time}",
      "limit"        => 20,
      "access_token" => token,
    })["data"] || []

    convos.each do |conv|
      msgs = (conv.dig("messages", "data") || []).sort_by { |m| m["created_time"].to_s }
      last = msgs.last
      next if last.nil?
      # l'ultimo messaggio deve essere dell'UTENTE, non nostro
      next if last.dig("from", "id").to_s == env["IG_USER_ID"].to_s
      sender = last.dig("from", "id").to_s
      next if sender.empty?
      next if state["dm_replied"][sender]

      rule = cfg["rules"].find { |r| matches?(last["message"], r["keywords"]) }
      next unless rule

      begin
        api_post("#{env['IG_USER_ID']}/messages", {
          "recipient" => { "id" => sender },
          "message"   => { "text" => rule["dm_message"] },
        }, token)
        state["dm_replied"][sender] = { "ok" => true, "rule" => rule["keywords"].first }
        log "Risposta DM a #{sender} [#{rule['keywords'].first}] (messaggio: #{last['message'].to_s[0, 60].inspect})"
        handled += 1
      rescue => e
        log "ERRORE risposta DM a #{sender}: #{e.message}"
      end
      File.write(STATE, JSON.pretty_generate(state))
      sleep 1
    end
   rescue => e
    # I DM richiedono l'App Review di Meta (accesso avanzato a
    # instagram_manage_messages). Finché non c'è, l'endpoint conversations
    # risponde "capability" (#3): lo registriamo una volta e NON blocchiamo
    # il resto del bot (i commenti funzionano comunque).
    log "DM non disponibili (serve App Review Meta): #{e.message}"
   end
  end

  File.write(STATE, JSON.pretty_generate(state))
  log "Passaggio completato: #{handled} invii" if handled > 0
rescue => e
  log "ERRORE: #{e.message}"
  exit 1
end
